### PR TITLE
More efficient working with headers

### PR DESCRIPTION
**Pull request by plokhotnyuk**
Currently, it is quite expensive in terms of burned CPU and eaten memory bandwidth during parsing, validation
image
...and serialization
image

This PR reduces that overheads and improves working with headers.

### CODE REVIEW

**Comment by plokhotnyuk** [comment]
@adamw Could, please review this PR?

**Comment by kciesielski** [comment]
@plokhotnyuk Thank you! Performance improvements like this one are much appreciated, and we are definitely interested in subsequent PRs. Unfortunately, the core code has to compile for Scala 2.12 as well. Could you adjust your changes to this requirement?

**Comment by plokhotnyuk** [comment]
@kciesielski Sorry, I missed to check that `sbt +core/test` doesn't run tests across Scala versions.

I've force pushed updated commit with fixes that had tested locally by the following sequence of commands:
```
sbt ++2.12.18 core/test
sbt ++2.13.11 core/test
sbt ++3.3.0 core/test
```

**Comment by kciesielski** [comment]
@plokhotnyuk We use the `sbt-projectmatrix` plugin, which generates dynamic projects for each scala version. The `core` project is for Scala 2.13, `core2_12` for 2.12 and `core3` for Scala 3. The failed compilation happens on `core2_12/compile` (compile scope).

**Comment by plokhotnyuk** [comment]
@kciesielski Thanks, for the heads up! Now `core2_12/compile` compiles without any errors.

**Comment by kciesielski** [comment]
Awesome, thanks again!

File: core/src/main/scala/sttp/model/headers/AcceptEncoding.scala [hunk 1]
```diff
@@ -34,40 +34,28 @@ object AcceptEncoding {
...
-    str.trim.split(",").toList match {
-      case x :: tail if x.nonEmpty =>
-        val range = parsSingleEncoding(x)
-        processString(tail.mkString(","), range :: acc)
-      case Nil => List(parsSingleEncoding(str))
-      case _   => acc
-    }
-  }
+  private def processString(str: String): List[WeightedEncoding] =
+    str.trim.split(",").map(x => parsSingleEncoding(x.trim)).reverse.toList // TODO: do we really need `.reverse` here?
```
Comment (adamw) [inline core/src/main/scala/sttp/model/headers/AcceptEncoding.scala:38]: regarding the TODO - yeah it looks weird, why would we need to reverse; if anything, we want to preserve the original ordering?

Comment (plokhotnyuk) [reply core/src/main/scala/sttp/model/headers/AcceptEncoding.scala:38]: Originally reversing was introduced by recursive version of previous implementation.

So my question in the TODO message was about need of reversion for the result list.

Comment (adamw) [reply core/src/main/scala/sttp/model/headers/AcceptEncoding.scala:38]: There was another tail-recursion with an accumulator which reversed the list - I moved the `reverse` to the more logical place: https://github.com/softwaremill/sttp-model/commit/ea43d9684b576c794f804c4bc200200b21d216de

Thanks for the PR, btw :)

**Pull request merged**